Sources/CacheAdvance/EncodableMessage.swift

```diff
@@ -0,0 +1,52 @@
+//
+//  Created by Dan Federman on 11/9/19.
+//  Copyright © 2019 Dan Federman.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// A struct that encodes a message of type T into data.
+/// A message is encoded with the following format:
+/// `[messageSize][data]`
+/// -  `messageSize` is a big-endian encoded `MessageSpan` of length `messageSpanStorageLength`.
+/// - `data` is length `messageSize`.
+struct EncodableMessage<T: Codable> {
+
+    // MARK: Initialization
+
+    /// Initializes an encoded message from the raw, codable source.
+    /// - Parameters:
+    ///   - message: The messages to encode.
+    ///   - encoder: The encoder to use.
+    init(message: T, encoder: JSONEncoder) {
```

Comment:
Random nerd thought: I suspect that writing this as a binary property list is more space efficient.

Reply:
almost certainly. Though Apple's APIs for doing so are more prone to crashing (rather than `throw`ing). Worth investigating though. I'll add this as an issue after merge.

Reply:
I just did an investigation. It looks like the binary property list is much less space efficient than a straight JSON encoding.

The reason (I think, I'd need to investigate further) is that I store messages one at a time, and the overhead of storing a list per message outweighs whatever benefits the binary property list gives us.

The following statements resolve to `true`:
```
try PropertyListSerialization.data(fromPropertyList: [], format: .binary, options: 0).count == 42
try JSONEncoder().encode(Data()).count == 2
```

Reply:
@dfed this makes some sense to me, since property lists have a header section and more overhead than JSON. What happens if we write a meaningful amount of data, eg 50 dictionaries with a few keys each?

Reply:
Surprisingly, encoding the entire `lorumIpsumMessages` array had the same result. Property list serialization is larger than json encoding.

```
try PropertyListSerialization.data(fromPropertyList: lorumIpsumMessages, format: .binary, options: 0).count == 7463
try JSONEncoder().encode(lorumIpsumMessages).count == 6822
```

I wonder if the JSONEncoder starts losing ground if each message has multiple fields. Right now we're only testing strings.

Reply:
Wow, good learnings here. Not what I expected. Thanks for checking!

Reply:
Just tried encoding an absurdly large dictionary, with each entry having hundreds of key:value pairs, and JSONEncoder still won out on size.

Reply:
Not what I expected either! Thanks for prompting me to check 🙂

Reply:
This likely won’t matter but I do want to point out there is a cleaner way to do this using `Codable`
https://developer.apple.com/documentation/foundation/propertylistencoder

Reply:
Good point! Just tried it, and:
```
try PropertyListEncoder().encode(lorumIpsumMessages).count == PropertyListSerialization.data(fromPropertyList: lorumIpsumMessages, format: .binary, options: 0).count
```